### PR TITLE
Test queryable built-in role synchronization

### LINKED ISSUE #118806: [CI] QueryableReservedRolesIT testDeletingAndCreatingSecurityIndexTriggersSynchronization failing
**Build Scans:**
- [elasticsearch-periodic #5427 / concurrent-search-tests](https://gradle-enterprise.elastic.co/s/tc5j5vaw7gyhm)
- [elasticsearch-periodic-platform-support #5415 / ubuntu-1804_platform-support-unix](https://gradle-enterprise.elastic.co/s/vxvykrj7r5rfa)
- [elasticsearch-periodic-platform-support #5415 / oraclelinux-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/vf3dqbf6ep5fa)
- [elasticsearch-periodic-platform-support #5415 / rhel-7_platform-support-unix](https://gradle-enterprise.elastic.co/s/ocqpzhsg7rwhw)
- [elasticsearch-pull-request #47015 / part-4](https://gradle-enterprise.elastic.co/s/lwap4q64cc7qo)
- [elasticsearch-periodic #5427 / graalvm-ce17_checkpart4_java-matrix](https://gradle-enterprise.elastic.co/s/gjrb45gwrqf44)
- [elasticsearch-periodic #5427 / openjdk22_checkpart4_java-matrix](https://gradle-enterprise.elastic.co/s/r225hapiamrke)
- [elasticsearch-periodic #5427 / openjdk17_checkpart4_java-matrix](https://gradle-enterprise.elastic.co/s/nug5ymtlig7wg)
- [elasticsearch-pull-request #47015 / part-4](https://gradle-enterprise.elastic.co/s/lyzkbuarl5cjc)
- [elasticsearch-periodic #5427 / openjdk21_checkpart4_java-matrix](https://gradle-enterprise.elastic.co/s/7zqp7amrr3vro)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:security:qa:security-basic:javaRestTest" --tests "org.elasticsearch.xpack.security.QueryableReservedRolesIT.testDeletingAndCreatingSecurityIndexTriggersSynchronization" -Dtests.seed=5E39A37157649705 -Dtests.locale=kea-CV -Dtests.timezone=America/Boise -Druntime.java=23
```

**Applicable branches:**
8.x

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.security.QueryableReservedRolesIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testDeletingAndCreatingSecurityIndexTriggersSynchronization),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: is <3>
     but: was <0>
```

**Issue Reasons:**
- [8.x] 4 consecutive failures in test testDeletingAndCreatingSecurityIndexTriggersSynchronization
- [8.x] 2 consecutive failures in step rhel-7_platform-support-unix
- [8.x] 2 consecutive failures in step openjdk17_checkpart4_java-matrix
- [8.x] 23 failures in test testDeletingAndCreatingSecurityIndexTriggersSynchronization (37.7% fail rate in 61 executions)
- [8.x] 2 failures in step rhel-7_platform-support-unix (100.0% fail rate in 2 executions)
- [8.x] 4 failures in step part-4 (33.3% fail rate in 12 executions)
- [8.x] 2 failures in step openjdk17_checkpart4_java-matrix (100.0% fail rate in 2 executions)
- [8.x] 3 failures in step part4 (27.3% fail rate in 11 executions)
- [8.x] 2 failures in pipeline elasticsearch-periodic (100.0% fail rate in 2 executions)
- [8.x] 2 failures in pipeline elasticsearch-periodic-platform-support (100.0% fail rate in 2 executions)
- [8.x] 3 failures in pipeline elasticsearch-pull-request (27.3% fail rate in 11 executions)
- [8.x] 3 failures in pipeline elasticsearch-intake (27.3% fail rate in 11 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### PR DESCRIPTION
**Pull request by slobodanadamovic**
Adds more tests for built-in roles synchronization, and fixes a bug where `synchronizationInProgress` hasn't been reset properly.

Resolves https://github.com/elastic/elasticsearch/issues/118806

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java [hunk 1]
```diff
@@ -201,18 +201,18 @@ public void clusterChanged(ClusterChangedEvent event) {
 
     private void syncBuiltInRoles(final QueryableBuiltInRoles roles) {
         if (synchronizationInProgress.compareAndSet(false, true)) {
-            final Map<String, String> indexedRolesDigests = readIndexedBuiltInRolesDigests(clusterService.state());
-            if (roles.rolesDigest().equals(indexedRolesDigests)) {
-                logger.debug("Security index already contains the latest built-in roles indexed, skipping synchronization");
-                return;
```
Comment (slobodanadamovic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java:207]: This is a bug fix where this method would return without calling `synchronizationInProgress.set(false);`.

**Review by n1v0lg — APPROVED** [review]
LGTM. Good catch! A few comments but no need for a re-review

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java [hunk 2]
```diff
@@ -201,17 +201,18 @@ public void clusterChanged(ClusterChangedEvent event) {
...
-                final Map<String, String> indexedRolesDigests = readIndexedBuiltInRolesDigests(clusterService.state());
-                if (roles.rolesDigest().equals(indexedRolesDigests)) {
-                    logger.debug("Security index already contains the latest built-in roles indexed, skipping roles synchronization");
-                } else {
-                    executor.execute(() -> doSyncBuiltinRoles(indexedRolesDigests, roles, ActionListener.wrap(v -> {
-                        logger.info("Successfully synced [" + roles.roleDescriptors().size() + "] built-in roles to .security index");
-                    }, QueryableBuiltInRolesSynchronizer::handleException)));
-                }
-            } finally {
+            final Map<String, String> indexedRolesDigests = readIndexedBuiltInRolesDigests(clusterService.state());
```
[the remark this answers is not in the document]
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java:204]: Absolutely, I'll wrap it with `try-catch` block. I don't see other option.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizerTests.java [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+        return nodes(randomIntBetween(1, 3), false).build();
+    }
+
+    private static DiscoveryNodes noDataNodes() {
+        return nodes(0, true).build();
+    }
+
+    private DiscoveryNodes mixedVersionNodes() {
+        VersionInformation oldVersion = new VersionInformation(
+            VersionUtils.randomCompatibleVersion(random(), VersionUtils.getPreviousVersion()),
```
[the remark this answers is not in the document]
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizerTests.java:422]: Good catch!

**Pull request merged**